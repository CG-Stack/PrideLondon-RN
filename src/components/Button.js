// @flow
import React from "react";
import { StyleSheet, Linking } from "react-native";
import Touchable from "./Touchable";
import Text from "./Text";
import {
  eventDetailsBuyButtonColor,
  eventDetailsBuyButtonTextColor,
  buttonPrimaryBackground,
  buttonPrimaryTextColor,
  greyishBrownColor,
  disabledButtonBgColor
} from "../constants/colors";

type ButtonType = "primary";

type Props = {
  text: string,
  url?: string,
  type?: ButtonType,
  onPress?: () => void,
  disabled?: boolean
};

export const handlePress = (onPress: ?() => void, url: ?string) => {
  if (onPress) {
    onPress();
  }

  if (url && url.length > 0) {
    Linking.openURL(url);
  }
};

const Button = ({ text, url, onPress, type = "primary", disabled }: Props) => (
  <Touchable
    style={[styles.button, styles[type], disabled ? styles.disabledButton : {}]}
    onPress={() => handlePress(onPress, url)}
    disabled={disabled}
  >
    <Text
      type="h2"
      style={[
        styles.text,
        styles[`${type}Text`],
        disabled ? styles.disabledText : {}
      ]}
    >
      {text}
    </Text>
  </Touchable>
);
Button.defaultProps = {
  url: "",
  type: "primary",
  onPress: () => {},
  disabled: false
};

const styles = StyleSheet.create({
  button: {
    backgroundColor: eventDetailsBuyButtonColor,
    borderRadius: 4,
    alignItems: "center",
    justifyContent: "center"
  },
  // eslint-disable-next-line react-native/no-unused-styles
  primary: {
    backgroundColor: buttonPrimaryBackground
  },
  // eslint-disable-next-line react-native/no-unused-styles
  primaryText: {
    color: buttonPrimaryTextColor
  },
  text: {
    color: eventDetailsBuyButtonTextColor
  },
  disabledButton: {
    backgroundColor: disabledButtonBgColor
  },
  disabledText: {
    color: greyishBrownColor
  }
});

export default Button;
