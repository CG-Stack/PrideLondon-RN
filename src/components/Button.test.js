// @flow
import React from "react";
import { Linking } from "react-native";
import { shallow } from "enzyme";
import Button, { handlePress } from "./Button";

let openURLSpy;
beforeEach(() => {
  openURLSpy = jest
    .spyOn(Linking, "openURL")
    .mockImplementation(() => Promise.resolve());
});

afterEach(() => {
  openURLSpy.mockRestore();
});

it("renders correctly", () => {
  const output = shallow(<Button text="a button" />);
  expect(output).toMatchSnapshot();
});

it("applies styles and props when disabled=true", () => {
  const output = shallow(<Button text="a button" disabled />);
  expect(output).toMatchSnapshot();
});

describe("#handlePress", () => {
  it("does nothing, if there is no url provided", () => {
    jest.mock("Linking", () => ({ openURL: jest.fn() }));
    handlePress();
    expect(Linking.openURL).not.toBeCalled();
  });

  it("opens the url, if there is one", () => {
    jest.mock("Linking", () => ({ openURL: jest.fn() }));
    handlePress(() => {}, "url");
    expect(Linking.openURL).toBeCalledWith("url");
  });
});
