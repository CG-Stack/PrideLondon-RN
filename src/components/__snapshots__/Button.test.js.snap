// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Touchable
  accessibilityComponentType="button"
  accessibilityTraits={
    Array [
      "button",
    ]
  }
  delayPressIn={50}
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#D8D8D8",
      "borderRadius": 4,
      "justifyContent": "center",
    }
  }
>
  <Text
    allowFontScaling={true}
    markdown={false}
    style={
      Object {
        "color": "#ffffff",
      }
    }
    type="text"
  >
    a button
  </Text>
</Touchable>
`;
